fabric-renderer-indigo/src/main/java/net/fabricmc/indigo/renderer/helper/BufferBuilderTransformHelper.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.indigo.renderer.helper;
+
+import com.google.common.collect.Sets;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+import net.minecraft.client.render.VertexFormat;
+import net.minecraft.client.render.VertexFormatElement;
+import net.minecraft.client.render.VertexFormats;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class BufferBuilderTransformHelper {
+	/**
+	 * Fast copying mode; used only if the vanilla format is an exact match.
+	 */
+	public static final int MODE_COPY_FAST = 0;
+
+	/**
+	 * Padded copying mode; used when the vanilla format is an exact match,
+	 * but includes additional data at the end. Will emit a warning.
+	 */
+	public static final int MODE_COPY_PADDED = 1;
+
+	/**
+	 * ShadersMod compatibility mode; as MODE_COPY_PADDED, but populates in
+	 * the correct normal values as provided by the mod.
+	 *
+	 * Assumes a format of [vertex, color, texture, lmap, normal], all in
+	 * their respective vanilla formats, plus any amount of additional data
+	 * afterwards.
+	 */
+	public static final int MODE_COPY_PADDED_SHADERSMOD = 2;
+
+	/**
+	 * Unsupported mode; an error will be emitted and no quads will be
+	 * pushed to the buffer builder.
+	 */
+	public static final int MODE_UNSUPPORTED = 3;
+
+	private static final Object2IntMap<VertexFormat> vertexFormatCache = new Object2IntOpenHashMap<>();
+	private static final Set<VertexFormat> errorEmittedFormats = Sets.newIdentityHashSet();
+	private static final Logger logger = LogManager.getLogger();
+
+	public static void emitUnsupportedError(VertexFormat format) {
+		if (errorEmittedFormats.add(format)) {
+			logger.error("[Indigo] Unsupported vertex format! " + format);
+		}
+	}
+
+	public static int getProcessingMode(VertexFormat format) {
+		return vertexFormatCache.computeIntIfAbsent(format, (f) -> {
```

Comment:
Strange as it seems, VertexFormats are mutable, and VertexFormat has a custom hashCode() implementation.

I don't know that we are obligated to support mods crazy enough to actually mutate VertexFormat instances at runtime, however it is probably worthwhile to avoid the hashCode()/map lookup overhead for the nominal case here.

One way to do that would be to save the last vertex format tested and do an == comparison before checking the hash map.

Or you could compare directly vs VertexFormats.POSITION_COLOR_UV_LMAP and VertexFormats.POSITION_COLOR_UV_NORMAL, in that order.  Those are static and you'll get an early out during block rendering, plus no memory write so it's probably the better option. 

That all said, it's not terrible as-is, and those improvements could come later if profiling shows them worthwhile.

Reply:
Also this needs to be thread-safe.  Static comparison as early out would not need to be.  Subsequent map look-up could be synchronized, or use ConcurrentHashMap to avoid synchronization at the cost of unboxing.